File: backend/services/payment_service.go
package services
import (
	"errors"
	"fmt"
	"time"
	"cronos-backend/config"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"github.com/google/uuid"
	"github.com/midtrans/midtrans-go"
	"github.com/midtrans/midtrans-go/snap"
)
type PaymentService struct {
	PaymentRepo *repositories.PaymentRepository
	OrderRepo   *repositories.OrderRepository
	UserRepo    *repositories.UserRepository
}
func NewPaymentService(paymentRepo *repositories.PaymentRepository, orderRepo *repositories.OrderRepository, userRepo *repositories.UserRepository) *PaymentService {
	return &PaymentService{
		PaymentRepo: paymentRepo,
		OrderRepo:   orderRepo,
		UserRepo:    userRepo,
	}
}
func (s *PaymentService) CreatePayment(orderID uuid.UUID) (*models.MidtransTransaction, error) {
	order, err := s.OrderRepo.FindByID(orderID)
	if err != nil {
		return nil, errors.New("order not found")
	}
	if order.Status != "pending" {
		return nil, errors.New("order is not in pending status")
	}
	existingPayment, _ := s.PaymentRepo.FindByOrderID(orderID)
	if existingPayment != nil && existingPayment.ID != uuid.Nil {
		if existingPayment.MidtransTransaction != nil {
			return existingPayment.MidtransTransaction, nil
		}
	}
	payment := &models.Payment{
		OrderID: orderID,
		Amount:  order.TotalAmount,
		Status:  "pending",
	}
	if err := s.PaymentRepo.Create(payment); err != nil {
		return nil, errors.New("failed to create payment")
	}
	var snapClient snap.Client
	if config.AppConfig.MidtransIsProduction {
		snapClient.New(config.AppConfig.MidtransServerKey, midtrans.Production)
	} else {
		snapClient.New(config.AppConfig.MidtransServerKey, midtrans.Sandbox)
	}
	midtransOrderID := fmt.Sprintf("CRONOS-%s", order.ID.String()[:8])
	user, _ := s.UserRepo.FindByID(order.UserID)
	snapReq := &snap.Request{
		TransactionDetails: midtrans.TransactionDetails{
			OrderID:  midtransOrderID,
			GrossAmt: int64(order.TotalAmount),
		},
		CustomerDetail: &midtrans.CustomerDetails{
			FName: user.Name,
			Email: user.Email,
			Phone: user.Phone,
		},
	}
	snapResp, snapErr := snapClient.CreateTransaction(snapReq)
	if snapErr != nil {
		return nil, fmt.Errorf("failed to create snap token: %v", snapErr)
	}
	mt := &models.MidtransTransaction{
		PaymentID:       payment.ID,
		OrderIDMidtrans: midtransOrderID,
		SnapToken:       snapResp.Token,
		SnapURL:         snapResp.RedirectURL,
	}
	if err := s.PaymentRepo.CreateMidtransTransaction(mt); err != nil {
		return nil, errors.New("failed to save midtrans transaction")
	}
	return mt, nil
}
func (s *PaymentService) HandleWebhook(notification map[string]interface{}) error {
	orderIDMidtrans, ok := notification["order_id"].(string)
	if !ok {
		return errors.New("invalid order_id in notification")
	}
	transactionStatus, _ := notification["transaction_status"].(string)
	paymentType, _ := notification["payment_type"].(string)
	fraudStatus, _ := notification["fraud_status"].(string)
	mt, err := s.PaymentRepo.FindMidtransByOrderID(orderIDMidtrans)
	if err != nil {
		return errors.New("midtrans transaction not found")
	}
	mt.TransactionStatus = transactionStatus
	mt.PaymentType = paymentType
	mt.FraudStatus = fraudStatus
	payment, err := s.PaymentRepo.FindByID(mt.PaymentID)
	if err != nil {
		return errors.New("payment not found")
	}
	if payment.Status == "paid" && (transactionStatus == "capture" || transactionStatus == "settlement") {
		s.PaymentRepo.UpdateMidtransTransaction(mt)
		return nil
	}
	switch transactionStatus {
	case "capture":
		if fraudStatus == "accept" {
			payment.Status = "paid"
			now := time.Now()
			payment.PaidAt = &now
			payment.Method = paymentType
			order, _ := s.OrderRepo.FindByID(payment.OrderID)
			order.Status = "paid"
			s.OrderRepo.Update(order)
			s.creditSellerBalance(order)
		}
	case "settlement":
		payment.Status = "paid"
		now := time.Now()
		payment.PaidAt = &now
		payment.Method = paymentType
		order, _ := s.OrderRepo.FindByID(payment.OrderID)
		order.Status = "paid"
		s.OrderRepo.Update(order)
		s.creditSellerBalance(order)
	case "deny", "cancel", "expire":
		payment.Status = "failed"
		if transactionStatus == "expire" {
			payment.Status = "expired"
		}
		order, _ := s.OrderRepo.FindByID(payment.OrderID)
		order.Status = "cancelled"
		s.OrderRepo.Update(order)
	case "pending":
		payment.Status = "pending"
	}
	s.PaymentRepo.Update(payment)
	s.PaymentRepo.UpdateMidtransTransaction(mt)
	return nil
}
func (s *PaymentService) creditSellerBalance(order *models.Order) {
	sellerAmounts := make(map[uuid.UUID]float64)
	for _, item := range order.OrderItems {
		var p models.Product
		s.PaymentRepo.DB.First(&p, item.ProductID)
		if p.ID != uuid.Nil {
			sellerAmounts[p.UserID] += item.Subtotal
		}
	}
	for sellerID, amount := range sellerAmounts {
		seller, _ := s.UserRepo.FindByID(sellerID)
		if seller != nil {
			seller.Balance += amount
			s.UserRepo.Update(seller)
		}
	}
}
func (s *PaymentService) GetPaymentByOrderID(orderID uuid.UUID) (*models.Payment, error) {
	return s.PaymentRepo.FindByOrderID(orderID)
}

File: backend-spring/src/main/java/com/cronos/service/PaymentService.java
package com.cronos.service;

import com.cronos.entity.*;
import com.cronos.repository.*;
import com.midtrans.Config;
import com.midtrans.ConfigFactory;
import com.midtrans.service.MidtransSnapApi;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService {
    private final PaymentRepository paymentRepo;
    private final MidtransTransactionRepository midtransRepo;
    private final OrderRepository orderRepo;
    private final UserRepository userRepo;
    private final ProductRepository productRepo;

    @Value("${midtrans.server-key}") private String serverKey;
    @Value("${midtrans.client-key}") private String clientKey;
    @Value("${midtrans.is-production}") private boolean isProduction;

    public MidtransTransaction createPayment(UUID orderId) {
        CronosOrder order = orderRepo.findById(orderId).orElseThrow(() -> new RuntimeException("order not found"));
        if (!"pending".equals(order.getStatus())) throw new RuntimeException("order is not in pending status");

        Optional<Payment> existingOpt = paymentRepo.findByOrderId(orderId);
        if (existingOpt.isPresent() && existingOpt.get().getMidtransTransaction() != null) {
            return existingOpt.get().getMidtransTransaction();
        }

        Payment payment = Payment.builder().orderId(orderId).amount(order.getTotalAmount()).status("pending").build();
        paymentRepo.save(payment);

        try {
            MidtransSnapApi snapApi = new ConfigFactory(new Config(serverKey, clientKey, isProduction)).getSnapApi();
            String midtransOrderId = "CRONOS-" + order.getId().toString().substring(0, 8);
            User user = userRepo.findById(order.getUserId()).orElse(null);

            Map<String, Object> params = new HashMap<>();
            Map<String, Object> txDetails = new HashMap<>();
            txDetails.put("order_id", midtransOrderId);
            txDetails.put("gross_amount", order.getTotalAmount().longValue());
            params.put("transaction_details", txDetails);

            if (user != null) {
                Map<String, String> custDetails = new HashMap<>();
                custDetails.put("first_name", user.getName());
                custDetails.put("email", user.getEmail());
                custDetails.put("phone", user.getPhone() != null ? user.getPhone() : "");
                params.put("customer_details", custDetails);
            }

            var snapResponse = snapApi.createTransaction(params);
            String snapToken = snapResponse.get("token") != null ? snapResponse.get("token").toString() : "";
            String snapUrl = snapResponse.get("redirect_url") != null ? snapResponse.get("redirect_url").toString() : "";

            MidtransTransaction mt = MidtransTransaction.builder()
                .paymentId(payment.getId()).orderIdMidtrans(midtransOrderId)
                .snapToken(snapToken).snapUrl(snapUrl).build();
            midtransRepo.save(mt);
            return mt;
        } catch (Exception e) {
            throw new RuntimeException("failed to create snap token: " + e.getMessage());
        }
    }

    @Transactional
    public void handleWebhook(Map<String, Object> notification) {
        String orderIdMidtrans = (String) notification.get("order_id");
        String transactionStatus = (String) notification.get("transaction_status");
        String paymentType = (String) notification.get("payment_type");
        String fraudStatus = (String) notification.get("fraud_status");

        MidtransTransaction mt = midtransRepo.findByOrderIdMidtrans(orderIdMidtrans)
                .orElseThrow(() -> new RuntimeException("midtrans transaction not found"));
        mt.setTransactionStatus(transactionStatus);
        mt.setPaymentType(paymentType);
        mt.setFraudStatus(fraudStatus);

        Payment payment = paymentRepo.findById(mt.getPaymentId()).orElseThrow(() -> new RuntimeException("payment not found"));

        if ("paid".equals(payment.getStatus()) && ("capture".equals(transactionStatus) || "settlement".equals(transactionStatus))) {
            midtransRepo.save(mt);
            return;
        }

        switch (transactionStatus) {
            case "capture" -> {
                if ("accept".equals(fraudStatus)) {
                    payment.setStatus("paid"); payment.setPaidAt(LocalDateTime.now()); payment.setMethod(paymentType);
                    CronosOrder order = orderRepo.findById(payment.getOrderId()).get();
                    order.setStatus("paid"); orderRepo.save(order); creditSellerBalance(order);
                }
            }
            case "settlement" -> {
                payment.setStatus("paid"); payment.setPaidAt(LocalDateTime.now()); payment.setMethod(paymentType);
                CronosOrder order = orderRepo.findById(payment.getOrderId()).get();
                order.setStatus("paid"); orderRepo.save(order); creditSellerBalance(order);
            }
            case "deny", "cancel", "expire" -> {
                payment.setStatus("expire".equals(transactionStatus) ? "expired" : "failed");
                CronosOrder order = orderRepo.findById(payment.getOrderId()).get();
                order.setStatus("cancelled"); orderRepo.save(order);
            }
            case "pending" -> payment.setStatus("pending");
        }
        paymentRepo.save(payment);
        midtransRepo.save(mt);
    }

    private void creditSellerBalance(CronosOrder order) {
        if (order.getOrderItems() == null) return;
        Map<UUID, BigDecimal> sellerAmounts = new HashMap<>();
        for (OrderItem item : order.getOrderItems()) {
            productRepo.findById(item.getProductId()).ifPresent(p -> sellerAmounts.merge(p.getUserId(), item.getSubtotal(), BigDecimal::add));
        }
        sellerAmounts.forEach((sellerId, amount) -> userRepo.findById(sellerId).ifPresent(seller -> {
            seller.setBalance(seller.getBalance().add(amount));
            userRepo.save(seller);
        }));
    }

    public Payment getPaymentByOrderId(UUID orderId) { return paymentRepo.findByOrderId(orderId).orElseThrow(() -> new RuntimeException("payment not found")); }
}
